src/protocol/resp/src/parse/parser.rs

```diff
@@ -0,0 +1,434 @@
+// Copyright 2023 Pelikan Foundation LLC.
+// Licensed under the Apache License, Version 2.0
+// http://www.apache.org/licenses/LICENSE-2.0
+
+use std::str::FromStr;
+
+use crate::parse::{ParseError, ParseResult};
+#[derive(Clone)]
+pub struct Parser<'a> {
+    data: &'a [u8],
+}
+
+impl<'a> Parser<'a> {
+    pub fn new(data: &'a [u8]) -> Self {
+        Self { data }
+    }
+
+    pub fn remaining(&self) -> &'a [u8] {
+        self.data
+    }
+
+    /// Fetch the next byte without advancing the parser.
+    pub fn peek(&self) -> Option<u8> {
+        self.data.first().copied()
+    }
+
+    /// Helper function that ensures a parse operation is applied in its
+    /// entirety or not at all.
+    pub fn try_parse<F, R>(&mut self, func: F) -> ParseResult<'a, R>
+    where
+        F: FnOnce(&mut Self) -> ParseResult<'a, R>,
+    {
+        let mut copy = self.clone();
```

Comment:
presumably cloning a reference is cheap, any reason that this couldn't/shouldn't be `Copy`? I could see that perhaps forcing a backwards incompatible change in the future, if we had to drop `Copy`

Reply:
I originally made this copy. However, I ran into some issues around it being copied when I didn't want it to so I removed the copy derive. Normally I would have it derive copy.